endpoints/openrtb2/auction.go

```diff
@@ -1277,25 +1277,73 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
 	}
 
 	// Check Universal User ID
-	for eidIndex, eid := range req.User.EIDs {
-		if eid.Source == "" {
-			return append(errL, fmt.Errorf("request.user.eids[%d] missing required field: \"source\"", eidIndex))
-		}
+	if len(req.User.EIDs) > 0 {
+
+		validEids, eidErrors := validateEIDs(req.User.EIDs)
 
-		if len(eid.UIDs) == 0 {
-			return append(errL, fmt.Errorf("request.user.eids[%d].uids must contain at least one element or be undefined", eidIndex))
+		if len(eidErrors) > 0 {
+			errL = append(errL, eidErrors...)
 		}
+		for eidIndex, eid := range validEids {
+			if eid.Source == "" {
+				return append(errL, fmt.Errorf("request.user.eids[%d] missing required field: \"source\"", eidIndex))
+			}
 
-		for uidIndex, uid := range eid.UIDs {
-			if uid.ID == "" {
-				return append(errL, fmt.Errorf("request.user.eids[%d].uids[%d] missing required field: \"id\"", eidIndex, uidIndex))
+			if len(eid.UIDs) == 0 {
+				return append(errL, fmt.Errorf("request.user.eids[%d].uids must contain at least one element or be undefined", eidIndex))
+			}
+
+			for uidIndex, uid := range eid.UIDs {
+				if uid.ID == "" {
+					return append(errL, fmt.Errorf("request.user.eids[%d].uids[%d] missing required field: \"id\"", eidIndex, uidIndex))
+				}
 			}
 		}
 	}
 
 	return errL
 }
 
+func validateEIDs(eids []openrtb2.EID) ([]openrtb2.EID, []error) {
+	var errorsList []error
+	validEIDs := make([]openrtb2.EID, 0, len(eids))
+
+	for eidIndex, eid := range eids {
+		validUIDs, uidErrors := validateUIDs(eid.UIDs)
+		errorsList = append(errorsList, uidErrors...)
+
+		if len(validUIDs) > 0 {
+			eid.UIDs = validUIDs
+			validEIDs = append(validEIDs, eid)
+		} else {
+			errorsList = append(errorsList, &errortypes.Warning{
+				Message:     fmt.Sprintf("request.user.eids[%d] (source: %s) contains only empty uids and is removed from the request", eidIndex, eid.Source),
+				WarningCode: errortypes.InvalidUserEIDsWarningCode,
+			})
+		}
+	}
+
+	return validEIDs, errorsList
+}
+
+func validateUIDs(uids []openrtb2.UID) ([]openrtb2.UID, []error) {
+	var validUIDs []openrtb2.UID
+	var uidErrors []error
+
+	for uidIndex, uid := range uids {
+		if uid.ID != "" {
+			validUIDs = append(validUIDs, uid)
+		} else {
+			uidErrors = append(uidErrors, &errortypes.Warning{
+				Message:     fmt.Sprintf("request.user.eids.uids[%d] contains empty ids and is removed from the request", uidIndex),
+				WarningCode: errortypes.InvalidUserUIDsWarningCode,
```

Comment:
I used the example of user.EIDs from the issue description: 
```json
"eids": [
			{
				"source": "uidapi.com",
				"uids": [
					{"id": ""},
					{"id": "AgAACYa0cG2jcVcM42xG"}
				]
			},
			{
				"source": "liveramp.com",
				"uids": [
					{"id": "AgAACYa0cG2jcVcM42xG"}
				]
			},
			{
				"source": "yahoo.com",
				"uids": [
					{"id": ""},
					{"id": ""}
				]
			}
		]
```
As a result I received a list of errors that looked like this: 
```
request.user.eids.uids[0] contains empty ids and is removed from the request
request.user.eids.uids[0] contains empty ids and is removed from the request
request.user.eids.uids[1] contains empty ids and is removed from the request
request.user.eids[2] (source: yahoo.com) contains only empty uids and is removed from the request
```

In order to avoid confusion and duplicates may you please pass eidIndex to `validateUIDs`:
Lines 1311 and 1312:
```Golang
for eidIndex, eid := range eids {
		validUIDs, uidErrors := validateUIDs(eid.UIDs, eidIndex)
```
and then add eidIndex to error message like this: 
Line 1338:
```Golang
Message:     fmt.Sprintf("request.user.eids[%d].uids[%d] contains empty ids and is removed from the request", eidIndex, uidIndex),
```

This will give error list with no duplicate error messages: 
```
request.user.eids[0].uids[0] contains empty ids and is removed from the request
request.user.eids[2].uids[0] contains empty ids and is removed from the request
request.user.eids[2].uids[1] contains empty ids and is removed from the request
request.user.eids[2] (source: yahoo.com) contains only empty uids and is removed from the request
```